site/content/extend/plugins/developer-workflow.md

```diff
@@ -0,0 +1,66 @@
+- Monitor webhook requests coming in with ngrok's request inspector. Visit [http://localhost:4040](http://localhost:4040) once you have your tunnel open. You can analyze the contents of the HTTP request from the external service, and the response from your plugin.
+
+If you are using a free `ngrok` account, the URL given by the output of the `ngrok http` command will be different each time you run the command. As a result, you will need to adjust the webhook URL on Mattermost's side and the external service's side (i.e. GitHub) each time you run the command.
+
+With this setup, many integrations require you to be logged into Mattermost using your ngrok URL. After logging into your `ngrok` URL pointed to your Mattermost server, you can in most cases continue using your `localhost` address in your browser for quicker network requests to your server. If you receive an error like `unauthorized` or `enable third-party cookies` when connecting to an external service, make sure you are logged into your `ngrok` URL in the same browser.
+
+### Debugging server-side plugins using `delve`
+
+Using the `make debug-plugin` command will allow you to use a debugger and step through the plugin's server code. A [delve](https://github.com/go-delve/delve) process will be created and attach to your plugin. You can then use an IDE or debug console to connect to the `delve` process. If you are using VSCode, you can use this `launch.json` configuration to connect.
+
+```json
+{
+    "name": "Attach remote",
+    "type": "go",
+    "request": "attach",
+    "mode": "remote",
+    "port": 2346,
+    "host": "127.0.0.1",
+    "apiVersion": 2
+}
+```
+
+If the debugger is paused for more than 5 seconds, the RPC connection with the server times out. The server cannot communicate with the plugin anymore, so the plugin then needs to be restarted.
+
+In order to be able to pause the debugger for more than 5 seconds, two modifications need to be done to the `mattermost-server` repository.
+
+1. The plugin health check job needs to be disabled. This can be done by commenting out the `go job.run()` line in [mattermost-server/plugin/environment.go](https://github.com/mattermost/mattermost-server/blob/master/plugin/environment.go#L510). Note that if your plugin crashes, you will need to restart it, using `make reset` for example. This will also kill any currently running `delve` process. If you want to continue debugging with `delve`, you will need to run `make debug-plugin` again after restarting the plugin.
+
+2. The `go-plugin`'s RPC client needs to be configured with a larger timeout duration. You can change the code at [mattermost-server/vendor/github.com/hashicorp/rpc_client.go](https://github.com/mattermost/mattermost-server/blob/bf03f391e635b0b9b129768cec5ea13c571744fa/vendor/github.com/hashicorp/go-plugin/rpc_client.go#L63) to increase the duration. Here's the change you can make to extend the timeout to 5 minutes:
```

Comment:
Should we edit the Mattermost server to make this configurable instead?

Reply:
@lieut-data Maybe, but it does need modifications to the vendor package either way. There's no way we can configure the timeout because `go-plugin` does not support a way to pass in a configuration for it. Maybe worth an upstream change?

Reply:
@lieut-data Any ideas on how to approach this? The `go-plugin` library needs to be modified directly. If we manually modify the existing version in the `vendor` folder, I'm concerned that the changes will be overwritten if the dependency is updated down the road.